cdap-ui/app/cdap/components/SecureKeys/index.tsx

```diff
@@ -14,166 +14,196 @@
+  Success = 'SUCCESS',
+  Failure = 'FAILURE',
+}
 
-  public componentDidMount() {
-    this.fetchSecureKeys();
+export const initialState = {
+  secureKeys: List([]),
+  secureKeyStatus: SecureKeyStatus.Normal,
+  editMode: false,
+  deleteMode: false,
+  activeKeyIndex: null,
+  loading: true,
+};
+
+export function reducer(state, action) {
+  switch (action.type) {
+    case 'SET_SECURE_KEYS':
+      return { ...state, secureKeys: action.secureKeys };
+    case 'SET_SECURE_KEY_STATUS':
+      return { ...state, secureKeyStatus: action.secureKeyStatus };
+    case 'SET_EDIT_MODE':
+      return { ...state, editMode: action.editMode };
+    case 'SET_DELETE_MODE':
+      return { ...state, deleteMode: action.deleteMode };
+    case 'SET_ACTIVE_KEY_INDEX':
+      return { ...state, activeKeyIndex: action.activeKeyIndex };
+    case 'SET_LOADING':
+      return { ...state, loading: action.loading };
+    default:
+      return state;
   }
+}
 
-  private fetchSecureKeys = () => {
-    const namespace = getCurrentNamespace();
-
-    MySecureKeyApi.list({ namespace }).subscribe((res) => {
-      this.setState({
-        secureKeys: res,
-      });
-    });
-  };
-
-  private handleChange = (property) => {
-    return (e) => {
-      // @ts-ignore
-      this.setState({
-        [property]: e.target.value,
-      });
-    };
+const styles = (): StyleRules => {
+  return {
+    content: {
+      padding: '50px',
+    },
+    loadingBox: {
+      width: '100%',
+      height: '100%',
+      display: 'flex',
+      alignItems: 'center',
+      justifyContent: 'center',
+    },
   };
+};
+
+interface ISecureKeysProps extends WithStyles<typeof styles> {}
+
+const SecureKeysView: React.FC<ISecureKeysProps> = ({ classes }) => {
+  const [state, dispatch] = React.useReducer(reducer, initialState);
+
+  const { secureKeyStatus, editMode, deleteMode, loading } = state;
+
+  const namespace = getCurrentNamespace();
+
+  // Observe `secureKeyStatus` with `useEffect` and forward the value to `secureKeyStatus$`
+  const secureKeyStatusSubject = React.useRef(new BehaviorSubject(secureKeyStatus));
+  React.useEffect(() => {
+    secureKeyStatusSubject.current.next(secureKeyStatus);
+  }, [secureKeyStatus]);
+  const secureKeyStatus$ = React.useMemo(() => secureKeyStatusSubject.current.asObservable(), [
+    secureKeyStatusSubject,
+  ]);
+
+  // Update the states with incoming secure keys
+  React.useEffect(() => {
+    // Whenever user adds/edits/deletes a secure key, securekeyStatus$ emits a new value.
+    // In such case, re-call MySecureKeyApi.list to reflect the changes
+    const secureKeys$ = secureKeyStatus$
+      .pipe(
+        distinctUntilChanged(),
+        flatMap((status) => {
+          return MySecureKeyApi.list({ namespace }).pipe(
+            mergeMap((keys: ISecureKeyState[]) => {
+              return forkJoin(
+                keys.map((k) =>
+                  MySecureKeyApi.getSecureData({ namespace, key: k.name }).pipe(
```

Comment:
you don't need to fetch the data. We don't want to give the ability to the user to view the content of the secure key. They should only be able to create the data or change it. Just like how it is with passwords.

Reply:
Resolved